File: src/routers/admin_link_profiles.py
"""Admin endpoints for per-inbound link profiles.

Why a separate router: profiles are a fourth admin-managed resource with
their own identity and validation. They describe extra client views of an
inbound and never change the rendered runtime config. No SQL and no profile
rules live here.
"""

from fastapi import APIRouter, Depends, HTTPException

import db
from models import LinkProfileIn, LinkProfileOut, LinkProfileUpdate
from routers.deps import conn
from services import link_profile_service

router = APIRouter(
    prefix="/api/admin/nodes/{node_id}/link-profiles",
    tags=["admin-link-profiles"],
)


async def _node_or_404(conn, node_id) -> dict:
    """Return the node row or raise 404.

    Why a helper: every profile endpoint starts with the same lookup, and a
    missing node must be a 404 everywhere.
    """
    node = await db.get_node(conn, node_id)
    if node is None:
        raise HTTPException(status_code=404, detail="node not found")
    return node


@router.get("", response_model=list[LinkProfileOut])
async def list_profiles(node_id: str, conn=Depends(conn)):
    """Return one node's link profiles in display order."""
    await _node_or_404(conn, node_id)
    return await link_profile_service.list_profiles(conn, node_id)


@router.post("", response_model=LinkProfileOut, status_code=201)
async def create_profile(node_id: str, payload: LinkProfileIn, conn=Depends(conn)):
    """Create a profile attached to one of the node's authored inbounds."""
    node = await _node_or_404(conn, node_id)
    errors = link_profile_service.validate_create(node, payload)
    if errors:
        raise HTTPException(status_code=422, detail=errors)
    if await link_profile_service.get_profile(conn, node_id, payload.id) is not None:
        raise HTTPException(status_code=409, detail="profile id already exists")
    return await link_profile_service.create_profile(conn, node_id, payload)


@router.get("/{profile_id}", response_model=LinkProfileOut)
async def get_profile(node_id: str, profile_id: str, conn=Depends(conn)):
    """Return one profile; 404 when the node or profile is missing."""
    await _node_or_404(conn, node_id)
    profile = await link_profile_service.get_profile(conn, node_id, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="profile not found")
    return profile


@router.put("/{profile_id}", response_model=LinkProfileOut)
async def update_profile(
    node_id: str, profile_id: str, payload: LinkProfileUpdate, conn=Depends(conn)
):
    """Partially update a profile; a missing profile is a 404."""
    node = await _node_or_404(conn, node_id)
    errors = link_profile_service.validate_update(node, payload)
    if errors:
        raise HTTPException(status_code=422, detail=errors)
    profile = await link_profile_service.update_profile(
        conn, node_id, profile_id, payload
    )
    if profile is None:
        raise HTTPException(status_code=404, detail="profile not found")
    return profile


@router.delete("/{profile_id}", status_code=204)
async def delete_profile(node_id: str, profile_id: str, conn=Depends(conn)):
    """Delete a profile; a missing profile is a 404."""
    await _node_or_404(conn, node_id)
    if not await link_profile_service.delete_profile(conn, node_id, profile_id):
        raise HTTPException(status_code=404, detail="profile not found")

File: src/routers/admin_link_profiles.ts
/**
 * Admin endpoints for per-inbound link profiles.
 *
 * WHY a separate router: profiles are a fourth admin-managed resource with
 * their own identity and validation. They describe extra client views of an
 * inbound and never change the rendered runtime config. No SQL and no
 * profile rules live here.
 */

import { Hono } from "hono";

import * as db from "../db";
import {
  ApiError,
  LinkProfileIn,
  LinkProfileUpdate,
  parseJsonBody,
} from "../models";
import * as linkProfileService from "../services/link_profile_service";

export const adminLinkProfiles = new Hono<{ Bindings: Env }>();

const PREFIX = "/api/admin/nodes/:node_id/link-profiles";

/**
 * WHAT: return the node row or throw 404.
 *
 * WHY a helper: every profile endpoint starts with the same lookup, and a
 * missing node must be a 404 everywhere.
 */
async function nodeOr404(env: Env, nodeId: string): Promise<db.NodeRow> {
  const node = await db.getNode(env.DB, nodeId);
  if (node === null) {
    throw new ApiError(404, "node not found");
  }
  return node;
}

/** WHAT: return one node's link profiles in display order. */
adminLinkProfiles.get(PREFIX, async (c) => {
  const nodeId = c.req.param("node_id");
  await nodeOr404(c.env, nodeId);
  return c.json(await linkProfileService.listProfiles(c.env.DB, nodeId));
});

/** WHAT: create a profile attached to one of the node's authored inbounds. */
adminLinkProfiles.post(PREFIX, async (c) => {
  const nodeId = c.req.param("node_id");
  // WHY the body is parsed before the node lookup: FastAPI resolved the
  // Pydantic model before the handler ran, so a schema-invalid body beat a
  // 404 even for a missing node.
  const payload = await parseJsonBody(c.req.raw, LinkProfileIn);
  const node = await nodeOr404(c.env, nodeId);
  const errors = linkProfileService.validateCreate(node, payload);
  if (errors.length > 0) {
    throw new ApiError(422, errors);
  }
  if (
    (await linkProfileService.getProfile(c.env.DB, nodeId, payload.id)) !== null
  ) {
    throw new ApiError(409, "profile id already exists");
  }
  return c.json(
    await linkProfileService.createProfile(c.env.DB, nodeId, payload),
    201,
  );
});

/** WHAT: return one profile; 404 when the node or profile is missing. */
adminLinkProfiles.get(`${PREFIX}/:profile_id`, async (c) => {
  const nodeId = c.req.param("node_id");
  await nodeOr404(c.env, nodeId);
  const profile = await linkProfileService.getProfile(
    c.env.DB,
    nodeId,
    c.req.param("profile_id"),
  );
  if (profile === null) {
    throw new ApiError(404, "profile not found");
  }
  return c.json(profile);
});

/** WHAT: partially update a profile; a missing profile is a 404. */
adminLinkProfiles.put(`${PREFIX}/:profile_id`, async (c) => {
  const nodeId = c.req.param("node_id");
  // WHY the body is parsed first: see the POST handler above.
  const payload = await parseJsonBody(c.req.raw, LinkProfileUpdate);
  const node = await nodeOr404(c.env, nodeId);
  const errors = linkProfileService.validateUpdate(node, payload);
  if (errors.length > 0) {
    throw new ApiError(422, errors);
  }
  const profile = await linkProfileService.updateProfile(
    c.env.DB,
    nodeId,
    c.req.param("profile_id"),
    payload,
  );
  if (profile === null) {
    throw new ApiError(404, "profile not found");
  }
  return c.json(profile);
});

/** WHAT: delete a profile; a missing profile is a 404. */
adminLinkProfiles.delete(`${PREFIX}/:profile_id`, async (c) => {
  const nodeId = c.req.param("node_id");
  await nodeOr404(c.env, nodeId);
  if (
    !(await linkProfileService.deleteProfile(
      c.env.DB,
      nodeId,
      c.req.param("profile_id"),
    ))
  ) {
    throw new ApiError(404, "profile not found");
  }
  return c.body(null, 204);
});
